libavformat/rtcenc.c

```diff
@@ -0,0 +1,2226 @@
+    if (SSL_CTX_set_tlsext_use_srtp(dtls_ctx, "SRTP_AES128_CM_SHA1_80")) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: SSL_CTX_set_tlsext_use_srtp failed\n");
+        ret = AVERROR(EINVAL);
+        goto end;
+    }
+
+end:
+    return ret;
+}
+
+/**
+ * After creating a DTLS context, initialize the DTLS SSL object.
+ */
+static av_cold int openssl_init_dtls_ssl(DTLSContext *ctx, SSL *dtls)
+{
+    int ret = 0;
+
+    /* Setup the callback for logging. */
+    SSL_set_ex_data(dtls, 0, ctx);
+    SSL_set_info_callback(dtls, openssl_on_info);
+
+    /* Set dtls fragment size */
+    SSL_set_options(dtls, SSL_OP_NO_QUERY_MTU);
+    /* Avoid dtls negotiate failed, limit the max size of DTLS fragment. */
+    SSL_set_mtu(dtls, ctx->pkt_size);
+
+#if OPENSSL_VERSION_NUMBER >= 0x10101000L /* OpenSSL 1.1.1 */
+    /* Set the callback for ARQ timer. */
+    DTLS_set_timer_cb(dtls, openssl_dtls_timer_cb);
+#endif
+
+    /* Setup DTLS as active, which is client role. */
+    SSL_set_connect_state(dtls);
+    SSL_set_max_send_fragment(dtls, ctx->pkt_size);
+
+    return ret;
+}
+
+/**
+ * Drives the SSL context by attempting to read packets to send from SSL, sending them
+ * over UDP, and then reading packets from UDP to feed back to SSL.
+ */
+static int openssl_drive_context(DTLSContext *ctx, SSL *dtls, BIO *bio_in, BIO *bio_out, int loop)
+{
+    int ret, i, j, r0, r1, req_size, res_size = 0;
+    uint8_t *data = NULL, req_ct = 0, req_ht = 0, res_ct = 0, res_ht = 0;
+    char buf[MAX_UDP_BUFFER_SIZE];
+    void *s1 = ctx->log_avcl;
+
+    /* Drive the SSL context by state change, arq or response messages. */
+    r0 = SSL_do_handshake(dtls);
+    r1 = SSL_get_error(dtls, r0);
+
+    /* Handshake successfully done */
+    if (r0 == 1) {
+        ctx->dtls_done_for_us = 1;
+        return 0;
+    }
+
+    /* Handshake failed with fatal error */
+    if (r0 < 0 && r1 != SSL_ERROR_WANT_READ) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: Start handshake failed, loop=%d, r0=%d, r1=%d\n", loop, r0, r1);
+        return AVERROR(EIO);
+    }
+
+    /* Fast retransmit the request util got response. */
+    for (i = 0; i <= ctx->dtls_arq_max && !res_size; i++) {
+        req_size = BIO_get_mem_data(bio_out, (char**)&data);
+        openssl_state_trace(ctx, data, req_size, 0, r0, r1);
+        ret = ffurl_write(ctx->udp_uc, data, req_size);
+        BIO_reset(bio_out);
+        req_ct = req_size > 0 ? data[0] : 0;
+        req_ht = req_size > 13 ? data[13] : 0;
+        if (ret < 0) {
+            av_log(s1, AV_LOG_ERROR, "DTLS: Send request failed, loop=%d, content=%d, handshake=%d, size=%d\n",
+                loop, req_ct, req_ht, req_size);
+            return ret;
+        }
+
+        /* Wait so that the server can process the request and no need ARQ then. */
+#if DTLS_PROCESSING_TIMEOUT > 0
+        av_usleep(DTLS_PROCESSING_TIMEOUT * 10000);
+#endif
+
+        for (j = 0; j <= DTLS_EAGAIN_RETRIES_MAX && !res_size; j++) {
```

Comment:
Derek Buitenhuis:

```text
> +        for (j = 0; j <= DTLS_EAGAIN_RETRIES_MAX && !res_size; j++) {
```

Does this not make DTLS_EAGAIN_RETRIES_MAX a misnomer (due to use of <=)?